k2/csrc/ragged_ops_inl.h

```diff
@@ -559,6 +560,59 @@ void ArgMaxPerSublist(Ragged<T> &src, T initial_value, Array1<int32_t> *dst) {
   }
 }
 
+template <typename T>
+void SegmentedExclusiveSum(Ragged<T> &src, Array1<T> *dst) {
+  ContextPtr c = GetContext(src, *dst);
+  int32_t dim = dst->Dim();
+  K2_CHECK_EQ(src.values.Dim(), dim);
+
+  // flags is similar to `tails` (see concepts in k2/csrc/utils)
+  // But it indicates `heads` here. The very first segment always
+  // starts at zero, so flags[0] is always 0.
+  Array1<uint32_t> flags(c, dim);
+  uint32_t *flags_data = flags.Data();
+
+  Array1<int32_t> row_splits = src.RowSplits(src.NumAxes() - 1);
+  Array1<int32_t> row_ids = src.RowIds(src.NumAxes() - 1);
+  const int32_t *row_splits_data = row_splits.Data();
+  const int32_t *row_ids_data = row_ids.Data();
+  K2_EVAL(
+      c, dim, set_flags, (int32_t idx01)->void {
+        int32_t idx0 = row_ids_data[idx01];
+        int32_t idx0x = row_splits_data[idx0];
+        int32_t idx0x_next = row_splits_data[idx0 + 1];
+        if (idx0x < idx0x_next) {
+          if (idx01 == idx0x)
+            flags_data[idx01] = idx01 == 0 ? 0 : 1;
+          else
+            flags_data[idx01] = 0;
+        }
+      });
+
+  if (c->GetDeviceType() == kCuda) {
+    // defined in k2/csrc/cudpp/cudpp.{h,cu}
+    SegmentedExclusiveSum(c, src.values.Data(), dim, flags_data, dst->Data());
+    return;
+  }
+
+  K2_CHECK_EQ(c->GetDeviceType(), kCpu);
+
+  const T *src_values_data = src.values.Data();
+  T *dst_data = dst->Data();
+
+  T sum = 0;
```

Comment:
It's better if we can put below lines into a if cpu branch, so that we can get the result for cpu in one iteration (with `flags` stuff).

Reply:
.. it effectively is a CPU branch, no?  
Might be clearer if it was in an else { }.  It did slightly confuse me.  But no big deal.